Deposits query

lehins opened the pull request:
# Description

This PR introduces a new `GetStakeDelegDeposits` query that is needed by cardano-cli to get to current deposits that delegators have left. As such it also bumps up versions in network

This is the first time I am adding a query, so please double check if I got all the versioning stuff right


**NOTE** - This PR is still missing a changelog entry. I need to figure out how this `scriv` stuff works, never used it before.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [x] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](https://github.com/input-output-hk/ouroboros-consensus/blob/main/docs/ReleaseProcess.md#adding-a-changelog-fragment).
    - [ ] Any changes in the Consensus API (every exposed function, type or module) that has changed its name, has been deleted, has been moved, or altered in some other significant way must leave behind a `DEPRECATED` warning that notifies downstream consumers. If deprecating a whole module, remember to add it to `./scripts/ci/check-stylish.sh` as otherwise `stylish-haskell` would un-deprecate it.
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

dnadales commented:
@lehins for adding the changelog fragment see [this section](https://github.com/input-output-hk/ouroboros-consensus/blob/main/docs/ReleaseProcess.md#adding-a-changelog-fragment). HTH!

dnadales reviewed (approved):
LGTM

amesgen reviewed (changes requested):
There are a few more places that need to be updated when adding a new protocol version, see e.g. c868d3244c55ed994daf5fdec974d012efe7d6be for reference (You don't *have* to introduce a new `CardanoNodeToClientVersion` as `CardanoNodeToClientVersion11` is not yet released).

CI is currently failing due to stylish-haskell failures, and there are also golden test failures which can be accepted by running the Shelley/Cardano test suite locally, as well as CDDL failures (see [here](https://github.com/input-output-hk/ouroboros-network/commit/c868d3244c55ed994daf5fdec974d012efe7d6be#diff-d57623b894b594cfb5feff5e47fdf2207fee488ed0306a88b758dc88236e630d) for how to fix them).

lehins commented on cabal.project:
```diff
@@ -21,7 +21,6 @@ index-state:
   , cardano-haskell-packages 2023-04-05T21:25:44Z
 
 packages: ./cardano-ping
-          ./ouroboros-network-testing
```
No, it was a duplicate. `ouroboros-network-testing` is listed twice

lehins commented on ouroboros-network-api/src/Ouroboros/Network/NodeToClient/Version.hs:
```diff
@@ -38,6 +38,8 @@ data NodeToClientVersion
     -- ^ added @GetPoolDistr, @GetPoolState, @GetSnapshots
     | NodeToClientV_15
     -- ^ enabled @CardanoNodeToClientVersion11@, i.e., Conway
+    | NodeToClientV_16
```
@amesgen I just noticed, thanks to you, that currently released version is NodeToClinetV_15

https://github.com/input-output-hk/ouroboros-network/blob/e4a222f8e0f5d55591cc14b062332a8a61e85724/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L531

Does this mean I should revert back addition of `NodeToClientV_16` and this becomes much smaller? All I need is addition of `ShelleyNodeToClientVersion7`, am I reading this right?

amesgen replied:
Right; yes, you can include the changes in `V_15` as Conway is unreleased :+1:

lehins replied:
`NodeToClientV_16` removed.

lehins commented on ouroboros-network-api/src/Ouroboros/Network/NodeToClient/Version.hs:
```diff
@@ -69,14 +72,15 @@ nodeToClientVersionCodec = CodecCBORTerm { encodeTerm, decodeTerm }
 
 nodeToClientVersionBit :: Int
-nodeToClientVersionBit = 15
+nodeToClientVersionBit = 16
```
Thank you for pointing this out. I've added a comment about this above the binding itself, to avoid future confusion. I also moved it inside the `were` clause from the top level to indicate that `nodeToClientVersionCodec` is the only function that uses it.

lehins commented:
@dnadales and @amesgen Thank you for your reviews.

> CI is currently failing due to stylish-haskell failures, 

Fixed.

> there are also golden test failures which can be accepted by running the Shelley/Cardano test suite locally, as well as CDDL failures.

I can't find any failures. It probably was due to the PR not being rebase on master. I've fixed some golden tests in my previous PR.

lehins commented:
There is something funny going on with sylish-haskell on cicero:
```
x86_64-linux > -module Ouroboros.Network.NodeToClient.Version (
x86_64-linux > - NodeToClientVersion (..)
x86_64-linux > +module Ouroboros.Network.NodeToClient.Version
x86_64-linux > + ( NodeToClientVersion (..)
```

This is not what `sylish-haskell` does for me locally. I am using the one from nix shell:
```
$ which stylish-haskell 
/nix/store/gl25c05xzmfmw0i658jpq1f1kw66f667-stylish-haskell-exe-stylish-haskell-0.14.4.0/bin/stylish-haskell
$ stylish-haskell -i ouroboros-network-api/src/Ouroboros/Network/NodeToClient/Version.hs
$ git diff
diff --git a/ouroboros-network-api/src/Ouroboros/Network/NodeToClient/Version.hs ... 
...
-module Ouroboros.Network.NodeToClient.Version
-  ( NodeToClientVersion (..)
+module Ouroboros.Network.NodeToClient.Version (
+    NodeToClientVersion (..)
...
```

amesgen commented:
@lehins Network uses a different configuration file for stylish-haskell, you can run
```
./scripts/ci/check-stylish-network.sh
```
to do that locally (there also is `./scripts/ci/check-stylish.sh` for Consensus).

lehins commented:
> Network uses a different configuration file for stylish-haskell

Yeah, I totally support you plan of splitting the two projects into separate repos.

lehins commented:
I think I made stylish-haskell happy enough.

Entries to changelog have been added.

However, I cannot decipher what is going on with cicero. Can someone give me a clue? Is it even related to this PR? If I had to guess the failure is unrelated, but I don't understand cicero output whatsoever.

amesgen reviewed (changes requested):
Thanks, looks good! The `CardanoNodeToClientVersion11` version still needs to be updated, using `ShelleyNodeToClientVersion7` instead of `6`.

https://github.com/input-output-hk/ouroboros-network/blob/e4a222f8e0f5d55591cc14b062332a8a61e85724/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L494-L509

> However, I cannot decipher what is going on with cicero. Can someone give me a clue? Is it even related to this PR? If I had to guess the failure is unrelated, but I don't understand cicero output whatsoever.

Yeah, it is sadly non-obvious to make instructive use of the Cicero output :sob: Usually, clicking on the "Loki" link and filtering for `x86_64-linux` or `ci (x86_64-linux)` works decently.

In this case, the issue is the missing golden test files I mentioned above. When you locally run the `ouroboros-consensus-shelley-test:test:test` suite locally, you should see the `ouroboros-consensus-shelley-test/test/golden/QueryVersion2/ShelleyNodeToClientVersion7/` being created, which you have to check in.

amesgen commented on ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:
```diff
@@ -505,6 +505,7 @@ pattern CardanoNodeToClientVersion11 =
       :* EraNodeToClientEnabled ShelleyNodeToClientVersion6
       :* EraNodeToClientEnabled ShelleyNodeToClientVersion6
       :* EraNodeToClientEnabled ShelleyNodeToClientVersion6
+      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
```
This does not typecheck, instead of adding a new component, update the existing versions from `6` to `7`, i.e.
```haskell
-- | The hard fork enabled, and the Shelley, Allegra, Mary, Alonzo, Babbage and
-- Conway eras enabled Using 'ShelleyNodeToClientVersion7' for the
-- Shelley-based eras, which enables new queries.
pattern CardanoNodeToClientVersion11 :: BlockNodeToClientVersion (CardanoBlock c)
pattern CardanoNodeToClientVersion11 =
    HardForkNodeToClientEnabled
      HardForkSpecificNodeToClientVersion2
      (  EraNodeToClientEnabled ByronNodeToClientVersion1
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* EraNodeToClientEnabled ShelleyNodeToClientVersion7
      :* Nil
      )
```

amesgen replied:
In your [latest force push](https://github.com/input-output-hk/ouroboros-network/compare/0112ca3ad92fdbd32cd17effbbe91bfe655660ca..53fb04bb324d17b7e942c3647ce417470e7f5383), you only edited the last Shelley number, and not all of them, as in the [snippet above](https://github.com/input-output-hk/ouroboros-network/pull/4507#discussion_r1171386510). That is not necessarily wrong, but it means that e.g. if there is a node release with Conway support, but the Conway HF has not yet happened, the deposit query will not work as we still have `ShelleyNodeToClientVersion6` for the penultimate era (Babbage).

lehins commented on ouroboros-network-api/src/Ouroboros/Network/NodeToClient/Version.hs:
```diff
@@ -37,7 +35,8 @@ data NodeToClientVersion
     | NodeToClientV_15
-    -- ^ enabled @CardanoNodeToClientVersion11@, i.e., Conway
+    -- ^ enabled @CardanoNodeToClientVersion11@, i.e., Conway and
+    -- @GetStakeDelegDeposits
```
Good catch. I've also fixed theses ones above: 
`@GetPoolDistr, @GetPoolState, @GetSnapshots` -> `@GetPoolDistr@, @GetPoolState@, @GetSnapshots@`

amesgen reviewed (approved):
:+1: on the versioning stuff, LGT`bors merge`d

lehins commented:
bors merge
